protocol/standard-l2-genesis.md

```diff
@@ -0,0 +1,185 @@
+the `xDomainMsgSender`. There is actually no need to set the value in storage during `initialize`, we could modify
+the semantics such that if its `address(0)` in storage, then return the default value, otherwise return the
+actual sender value. This should be safe since there is no way to be a sender from `address(0)`.
+
+Given this insight and the fact that there is reentrancy check on `relayMessage`, it should be safe to use transient
+storage without a call depth context. There is an [open PR](https://github.com/ethereum-optimism/optimism/pull/12356) to migrate to solc `0.8.25`.
+
+## Resource Usage
+
+<!-- What is the resource usage of the proposed solution?
+Does it consume a large amount of computational resources or time? -->
+
+The additional deposit gas is the only additional resource usage and its covered in the risks section
+at the bottom of this document.
+
+This approach expands the ABI of the `L1Block` contract, meaning that automatically generated solidity dispatcher
+will binary search over the possible function selectors, consuming a bit more gas. This is something that we could
+avoid by writing a custom dispatcher in a fallback function or by grinding for names of functions.
+
+# Alternatives Considered
+
+<!-- List out a short summary of each possible solution that was considered.
+Comparing the effort of each solution -->
+
+There is a long history of alternatives here.
+
+Another option would be to embed these config values directly into the client software's config and have the client
+software create these deposit txs rather than the smart contracts. This is less flexible but comes with the tradeoff
+of additional required rollup config and doesn't solve the problem for existing chains. Existing chains would need a way
+to source this config, it would likely need to be hardcoded in the binary and that isn't super scalable.
+
+# Risks & Uncertainties
+
+<!-- An overview of what could go wrong.
+Also any open questions that need more work to resolve. -->
+
+There is a concern that the sequencer can include transactions before these values are set on L2.
+If we define the `SystemConfig.startBlock` as the [first block to start derivation in](https://github.com/ethereum-optimism/optimism/blob/d05fb505809717282d5cee7264a09d26002a4ddd/op-node/cmd/genesis/cmd.go#L174C30-L174C40),
+which is set on `SystemConfig.initialize` and also the deposit transactions that set these values are
+sent in the same block, then we should have the guarantee that no user transactions are included before
+the deposit transactions.
+
+There is a concern around the max deposit gas limit being too small so that the `SystemConfig` cannot
+deposit all of these values, we should have logic that reverts if the deposit gas limit is too small
+in the `ResourceConfig` sanity check function. Since the `ResourceConfig` can be modified during
+`initialize`, its not that big of a deal and chain operators will see that they need to increase that
```

Comment:
> Since the `ResourceConfig` can be modified during `initialize`, its not that big of a deal

Is the idea that it can be increased in the event that the cost of the setConfig() calls becomes to large?

> and chain operators will see that they need to increase that

how do chain operators do that, and why do they need to?

Reply:
Yes that is the idea, it will revert during `initialize` if the deposit gas limit is too small, so we should make sure that the resource config is updated before any of the deposit txs are created, so that the chain operator can increase the deposit gas limit as part of the same initialize call. They may need to increase the deposit gas limit since deposits are issued as part of the call to `initialize`

Reply:
> so that the chain operator can increase the deposit gas limit as part of the same initialize call

In the context of the superchain though, it is the L1ProxyAdmin owner, not a chain operator, who will call `initialize` during an upgrade. AFAICT, no action would be required by a chain operator.

Reply:
I think this is sufficient clarification: dc28731d4ec4